pkg/algorithm/hived_algorithm_tester.go

```diff
@@ -0,0 +1,305 @@
+	h := tester.h
+	t := tester.t
+	defer func() {
+		if err := recover(); err != nil {
+			// record the panic
+			t.Logf("Panic detected for pod %v. Details: %v", podName, err)
+			tester.panicPodNames[podName] = true 
+		}
+	}()
+	pod := &core.Pod{
+			ObjectMeta: meta.ObjectMeta{
+				Name:        podName,
+				Namespace:   "test",
+				UID:         types.UID(podName),
+				Annotations: map[string]string{},
+			},
+	}
+	pod.Annotations[api.AnnotationKeyPodSchedulingSpec] = common.ToYaml(podGroupSchedulingRequest)
+	psr := h.Schedule(pod, tester.allNodes, phase)
+	if psr.PodBindInfo != nil {
+		allocatedPod := internal.NewBindingPod(pod, psr.PodBindInfo)
+		h.AddAllocatedPod(allocatedPod)
+		tester.podScheduleResult[podName] = psr
+		tester.pods[podName] = allocatedPod
+	} else if psr.PodPreemptInfo != nil {
+		for _, victimPod := range psr.PodPreemptInfo.VictimPods {
+			h.DeleteUnallocatedPod(victimPod)
+		}
+		tester.pods[podName] = pod
+		tester.podScheduleResult[podName] = psr
+	} else {
+		tester.pods[podName] = pod
+		tester.podScheduleResult[podName] = psr
+	}
+}
+
+func (tester *GenericHivedAlgorithmTester) DeallocatePod(podName string) {
+	h := tester.h
+	pod, ok := tester.pods[podName]
+	if ok {
+		h.DeleteUnallocatedPod(pod)
+		delete(tester.pods, podName);
+	} else {
+		panic("Cannot find pod " + podName)
+	}
+}
+
+func (tester *GenericHivedAlgorithmTester) AssertPodBindResult(podName string, expectedResult bindResult) {
+	t := tester.t
+	psr, ok := tester.podScheduleResult[podName]
+	if ok {
+		if psr.PodBindInfo == nil {
+			t.Errorf("AssertPodBindResult failed for pod %v: Cannot find valid PodBindInfo!", podName)
+		} else {
+			podBindInfo := *psr.PodBindInfo
+			if expectedResult.Node != podBindInfo.Node {
+				t.Errorf("AssertPodBindResult failed for pod %v: Expected node is %v but got node %v", podName, expectedResult.Node, podBindInfo.Node)
+			} else {
+				if !reflect.DeepEqual(expectedResult.LeafCellIsolation, podBindInfo.LeafCellIsolation) {
+					t.Errorf("AssertPodBindResult failed for pod %v: Expected LeafCellIsolation is %v but got %v", podName, 
+						expectedResult.LeafCellIsolation, podBindInfo.LeafCellIsolation)
+				} else {
+					t.Logf("AssertPodBindResult ok for pod %v.", podName)
+				}
+			}
+		}
+	} else {
+		t.Errorf("AssertPodBindResult failed for pod %v: Cannot find valid schedule result!", podName)
+	}
+}
+
+func (tester *GenericHivedAlgorithmTester) AssertPodPrememptResult(podName string, expectedResult preemptResult) {
+	t := tester.t
+	psr, ok := tester.podScheduleResult[podName]
+	if ok {
+		if psr.PodPreemptInfo == nil {
+			t.Errorf("AssertPodPrememptResult failed for pod %v: Cannot find valid PodPreemptInfo!", podName)
+		} else {
+			victimPodNames := make([]string, len(psr.PodPreemptInfo.VictimPods))
+			for _, pod := range(psr.PodPreemptInfo.VictimPods) {
+				victimPodNames = append(victimPodNames, pod.Name)
+			}
```

Comment:
it will make `victimPodNames` twice of `len`
```suggestion
			victimPodNames := []string{}
			for _, pod := range psr.PodPreemptInfo.VictimPods {
				victimPodNames = append(victimPodNames, pod.Name)
			}
```

Reply:
done